cartoframes/data/observatory/enrichment/enrichment_service.py

```diff
@@ -83,358 +66,56 @@ def _prepare_data(self, dataframe, geom_col):
-        raise EnrichmentError('variables parameter should be a Variable instance, a list or a str.')
-
-    if not isinstance(variables, Variable) and len(variables) < 1:
-        raise EnrichmentError('You should add at least one variable to be used in enrichment.')
-
-    if isinstance(variables, list) and len(variables) > MAX_VARIABLES_NUMBER:
-        raise EnrichmentError('The maximum number of variables to be used in enrichment is 50.')
-
-
-def _validate_bq_operations(variables, credentials):
-    dataset_ids = list(set([variable.dataset for variable in variables]))
+    @timelogger
+    def _upload_data(self, temp_table_name, geodataframe):
+        reduced_geodataframe = geodataframe[[_ENRICHMENT_ID, _GEOM_COLUMN]]
 
-    for dataset_id in dataset_ids:
-        dataset = Dataset.get(dataset_id)
-        geography = Geography.get(dataset.geography)
+        dataset = DODataset(auth_client=self.auth_client).name(temp_table_name) \
+            .column(_ENRICHMENT_ID, 'INT64') \
+            .column(_GEOM_COLUMN, 'GEOMETRY') \
+            .ttl_seconds(_TTL_IN_SECONDS)
+        dataset.create()
 
-        _is_subscribed(dataset, geography, credentials)
-        _is_available_in_bq(dataset, geography)
+        status = dataset.upload_dataframe(reduced_geodataframe, _GEOM_COLUMN)
 
+        if status not in ['success']:
+            raise EnrichmentError('Couldn\'t upload the dataframe to be enriched. The job hasn\'t finished successfuly')
 
-def _is_available_in_bq(dataset, geography):
-    if not dataset._is_available_in('bq'):
-        raise EnrichmentError("""
-            The Dataset '{}' is not ready for Enrichment. Please, contact us for more information.
-        """.format(dataset))
+        return dataset
 
-    if not geography._is_available_in('bq'):
-        raise EnrichmentError("""
-            The Geography '{}' is not ready for Enrichment. Please, contact us for more information.
-        """.format(geography))
+    @timelogger
+    def _execute_enrichment(self, dataset, temp_table_name, geom_type, variables, filters, aggregation):
+        output_name = '{}_result'.format(temp_table_name)
+        status = dataset.enrichment(geom_type=geom_type,
+                                    variables=variables,
+                                    filters=filters,
+                                    aggregation=aggregation,
+                                    output_name=output_name)
 
+        if status not in ['success']:
+            raise EnrichmentError('Couldn\'t enrich the dataframe. The job hasn\'t finished successfuly')
 
-def _is_subscribed(dataset, geography, credentials):
-    if not dataset._is_subscribed(credentials):
-        raise EnrichmentError("""
-            You are not subscribed to the Dataset '{}' yet. Please, use the subscribe method first.
-        """.format(dataset.id))
+        result = DODataset(auth_client=self.auth_client).name(output_name).download_stream()
```

Comment:
Same here using `self.do_dataset`

Reply:
We need a new one for each case since each one creates a different dataset in BQ. If not, rerunning a cell in cartoframes becomes in an error 
https://github.com/CartoDB/cartoframes/commit/9985832b24d09966e61a665e284fd0d79c5bc7de

Reply:
OK. Thanks for the information.

Reply:
it prevents https://app.clubhouse.io/cartoteam/story/61906/error-500-field-enrichment-id-already-exists-in-schema